src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs

```diff
@@ -224,8 +224,37 @@ private Expression BindCountNode(CountNode node)
                 countMethod = ExpressionHelperMethods.EnumerableCountGeneric.MakeGenericMethod(elementType);
             }
 
-            // call Count() method. 
-            countExpression = Expression.Call(null, countMethod, new[] { source });
+            MethodInfo whereMethod;
+            if (typeof(IQueryable).IsAssignableFrom(source.Type))
+            {
+                whereMethod = ExpressionHelperMethods.QueryableWhereGeneric.MakeGenericMethod(elementType);
+            }
+            else
+            {
+                whereMethod = ExpressionHelperMethods.EnumerableWhereGeneric.MakeGenericMethod(elementType);
+            }
+
+            // Bind the inner $filter clause within the $count segment.
+            // e.g Books?$filter=Authors/$count($filter=Id gt 1) gt 1
+            Expression filterExpression = null;
+            if (node.FilterClause != null)
```

Comment:
It seems we don't need the else {} clause.
```C#
if (node.FilterClause != null)
{
    .....
   source = Expression.Call(null, ....);
}

countExpression = Expression.Call(null, countMethod, new[] { source });
```

Reply:
#Resolved